### PR TITLE
Implements possibility to unsubscribe from email reports

### LINKED ISSUE #2197: Possibility to unsubscribe from Email reports (pdf, html, csv)
I received some emails with this feature request since they were not able to switch off the email reports (probably via the additional email-addresses field). And they were not even able to determine, who had set up this reports.

Possible solutions
- Unsubscribe Link in the Reporting eMails
- Provide information about the user who set up the report (e-Mail-Address, etc. - so the receiver can quickly get in contact with him)

Any thoughts on this?

### PR DESCRIPTION
**Pull request by sgiehl**
This is a POC for report unsubscriptions. Maybe someone could have a look if the implementation is fine that way.

Would add tests and some more doc blocks and so on after a first review...

fixes #2197

### CODE REVIEW

**Comment by mattab** [comment]
Thanks for the PR!  Feedback after a quick review:

* overall looks good to me, maybe @diosmosis or @tsteur have further thoughts?
* maybe we need 2 tokens in the unsubscribe URL to be secure, otherwise one could write a script looping over all possible token values and unsubscribe everyone (after a very long time)? for example a hash of the idreport could maybe do..
* instead of updating the subscriptions when sending emails, could we have the migration script do it once, and then have the other updateReport addReport etc. APIs update the list of subscriptions?
* left few other comments

**Comment by sgiehl** [comment]
@mattab  I'll implement a Nonce in the unsubscription form, that should help avoiding iterating automatically over possible tokens (which have up to 100 chars).

For sure we could also update the subscription when the reports are added/updated. I've Implemented it that way to ensure a subscription is created/available if an email is sent. Otherwise it might happen that for example third party plugins somehow manipulate reports without calling a method that updates the subscriptions.

**Comment by diosmosis** [comment]
Came up w/ one question that probably only @mattab can answer: Is it important for admins to see who unsubscribed?

**Comment by sgiehl** [comment]
@diosmosis  Thought about that as well. Will add an event triggered when someone unsubscribes from a report. That way we can at least show it in the Activity Log

**Comment by mattab** [comment]
It may be needed to keep trace of those who unsubscribed. Would it maybe work to have a column in the table "subscribed" or so and set it to 0 when not? This way we could have the data who has unsubscribed, which later we may show in the UI maybe. probably only showing this info in the activity log is required for now?

**Comment by sgiehl** [comment]
@mattab  I've pushed an update. The table now also includes two columns with timestamps for subscription and unsubscription date. When someone unsubscribes the unsubscription date will be set, and the token will be nulled. If the email is readded in the report this entry will be overwritten with a new one (with unset unsubscribe date). That way we could later show a list of emails that unsubscribed or maybe even block readding unsubscribed emails in report admin.

**Review by mattab — CHANGES_REQUESTED** [review]

**Comment by mattab** [comment]
Update looks great @sgiehl :+1:  Just left a few comments, let me know if these could be applied soon?

File: plugins/ScheduledReports/SubscriptionModel.php [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                $report['parameters']['emailMe'] = false;
+            }
+        }
+
+        if ($emailFound) {
+            Access::doAsSuperUser(function() use ($report) {
+                Request::processRequest('ScheduledReports.updateReport', [
+                    'idReport'     => $report['idreport'],
+                    'idSite'       => $report['idsite'],
+                    'description'  => $report['description'],
```
[the remark this answers is not in the document]
Comment (sgiehl) [reply plugins/ScheduledReports/SubscriptionModel.php]: yes. the parameters are all required for the update method. See https://github.com/matomo-org/matomo/blob/3.x-dev/plugins/ScheduledReports/API.php#L139

File: plugins/ScheduledReports/SubscriptionModel.php [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    public function getSubscription($token)
+    {
+        return $this->getDb()->fetchRow('SELECT * FROM ' . $this->table . ' WHERE token = ?', [$token]);
+    }
+
+    public function updateReportSubscriptions($idReport, $emails)
+    {
+        $availableSubscriptions = $this->getReportSubscriptions($idReport);
```
[the remark this answers is not in the document]
Comment (sgiehl) [reply plugins/ScheduledReports/SubscriptionModel.php]: no. the method is also used to initially set the subscriptions.

File: plugins/ScheduledReports/templates/unsubscribe.twig [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    {{ postEvent("Template.beforeTopBar", "login") }}
+    {{ postEvent("Template.beforeContent", "login") }}
+
+    {% include "_iframeBuster.twig" %}
+
+    <div id="notificationContainer">
+    </div>
+    <nav>
+        <div class="nav-wrapper">
+            <span id="logo" class="brand-logo center">
```
[the remark this answers is not in the document]
Comment (sgiehl) [reply plugins/ScheduledReports/templates/unsubscribe.twig]: we could extract it into another template and include it into both. but we also have similar implementations in all layout.twig files (morpheus, updater, ...)

File: plugins/ScheduledReports/ScheduledReports.php [hunk 4]
```diff
@@ -343,12 +344,39 @@ public function sendReport($reportType, $report, $contents, $filename, $prettyDa
...
 
+        $subscriptionModel = new SubscriptionModel();
+        $subscriptionModel->updateReportSubscriptions($report['idreport'], $emails);
+        $subscriptions = $subscriptionModel->getReportSubscriptions($report['idreport']);
+
+        $tokens = [];
+
+        foreach ($subscriptions as $subscription) {
+            $tokens[$subscription['email']] = $subscription['token'];
+        }
```
Comment (diosmosis) [inline plugins/ScheduledReports/ScheduledReports.php]: Think this could be shortened to `$tokens = array_column($subscriptions, 'token', 'email');`

Comment (diosmosis) [reply plugins/ScheduledReports/ScheduledReports.php]: Not really important, just fyi

File: plugins/ScheduledReports/SubscriptionModel.php [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        $email = $details['email'];
+
+        $report = Access::doAsSuperUser(function() use ($details) {
+            $reports = Request::processRequest('ScheduledReports.getReports', array(
+                'idReport'    => $details['idreport'],
+            ));
+            return reset($reports);
+        });
+
+        if (empty($report)) {
```
Comment (diosmosis) [inline plugins/ScheduledReports/SubscriptionModel.php]: In the case where there's no report for a subscription, should the subscription be removed?

Comment (diosmosis) [reply plugins/ScheduledReports/SubscriptionModel.php]: Note: think this is an edge case, so could put this in a new ticket if important

Comment (sgiehl) [reply plugins/ScheduledReports/SubscriptionModel.php]: changed

**Pull request merged**